components/net/filemanager_thread.rs

```diff
@@ -489,6 +512,103 @@ impl FileManagerStore {
+                format!("inline; {}",
+                    if charset.to_lowercase() == "utf-8" {
+                        format!("filename=\"{}\"", String::from_utf8(name.as_bytes().into()).unwrap())
+                    } else {
+                        format!("filename*=\"{}\"''{}", charset, http_percent_encode(name.as_bytes()))
+                    }
+                ).as_bytes()
+            ).unwrap()
+        );
+    }
+
+    fn fetch_blob_buf(&self, sender: &servo_channel::Sender<Data>,
+                    id: &Uuid, origin_in: &FileOrigin, rel_pos: RelativePos,
+                    check_url_validity: bool, response: &mut Response) -> Result<(), BlobURLStoreError> {
+        let mut bytes = vec![];
+        let file_impl = self.get_impl(id, origin_in, check_url_validity)?;
+        match file_impl {
+            FileImpl::Memory(buf) => {
+                let range = rel_pos.to_abs_range(buf.size as usize);
+                let len = range.len() as u64;
+
+                self.set_headers(response,
+                                 len,
+                                 buf.type_string.parse().unwrap_or(mime::TEXT_PLAIN),
+                                 /* filename */ None);
+
+                bytes.extend_from_slice(buf.bytes.index(range));
+
+                let _ = sender.send(Data::Payload(bytes));
+                let _ = sender.send(Data::Done);
+
+                Ok(())
+            },
+            FileImpl::MetaDataOnly(metadata) => {
+                /* XXX: Snapshot state check (optional) https://w3c.github.io/FileAPI/#snapshot-state.
+                        Concretely, here we create another file, and this file might not
+                        has the same underlying file state (meta-info plus content) as the time
+                        create_entry is called.
+                */
+
+                let mut file = File::open(&metadata.path)
+                                   .map_err(|e| BlobURLStoreError::External(e.to_string()))?;
+
+                let range = rel_pos.to_abs_range(metadata.size as usize);
+                let range_start = range.start as u64;
+                let seeked_start = file.seek(SeekFrom::Start(range_start))
+                                       .map_err(|e| BlobURLStoreError::External(e.to_string()))?;
+                if seeked_start != range_start {
+                    return Err(BlobURLStoreError::InvalidEntry)
+                }
+
+                let filename = metadata
+                    .path
+                    .file_name()
+                    .and_then(|osstr| osstr.to_str())
+                    .map(|s| s.to_string());
+
+                self.set_headers(response,
+                                 metadata.size,
+                                 guess_mime_type_opt(metadata.path).unwrap_or(mime::TEXT_PLAIN),
+                                 filename);
+
+                let body = response.body.clone();
+                let sender = sender.clone();
```

Comment:
Why do we need to clone the sender?

Reply:
Because we need to use it again [here](https://github.com/servo/servo/pull/22134/files#diff-9f3e543e8b941fbfee0d057ce7ff9880R35).